language/vm/src/deserializer.rs

```diff
@@ -42,6 +38,111 @@ impl CompiledModuleMut {
     }
 }
 
+mod versioned_data {
+    use crate::{errors::*, file_format_common::*};
+    use move_core_types::vm_status::StatusCode;
+    use std::io::{Cursor, Read};
+    pub(super) struct VersionedBinary<'a> {
+        version: u32,
+        binary: &'a [u8],
+    }
+
+    pub(super) struct VersionedCursor<'a> {
+        version: u32,
+        cursor: Cursor<&'a [u8]>,
+    }
+
+    pub const MAX_VERSION: u32 = 1u32;
+
+    impl<'a> VersionedCursor<'a> {
+        /// Verifies the correctness of the "static" part of the binary's header.
+        /// If valid, returns a cursor to the binary
+        pub fn new(binary: &'a [u8]) -> BinaryLoaderResult<Self> {
+            let mut cursor = Cursor::<&'a [u8]>::new(binary);
+            let mut magic = [0u8; BinaryConstants::DIEM_MAGIC_SIZE];
+            if let Ok(count) = cursor.read(&mut magic) {
+                if count != BinaryConstants::DIEM_MAGIC_SIZE || magic != BinaryConstants::DIEM_MAGIC
+                {
+                    return Err(PartialVMError::new(StatusCode::BAD_MAGIC));
+                }
+            } else {
+                return Err(PartialVMError::new(StatusCode::MALFORMED)
+                    .with_message("Bad binary header".to_string()));
+            }
+            let version = match read_u32(&mut cursor) {
+                Ok(v) => v,
+                Err(_) => {
+                    return Err(PartialVMError::new(StatusCode::MALFORMED)
+                        .with_message("Bad binary header".to_string()));
+                }
+            };
+            if version == 0 || version > MAX_VERSION {
+                return Err(PartialVMError::new(StatusCode::UNKNOWN_VERSION));
+            }
+            Ok(VersionedCursor { version, cursor })
+        }
```

Comment:
Maybe put this into a `fn VersionedBinary::new`? It feels weird to me that the binary is derived from a cursor, rather than the other way around.